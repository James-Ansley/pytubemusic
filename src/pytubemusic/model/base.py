from datetime import timedelta
from pathlib import PurePath
from pprint import PrettyPrinter

from pydantic import BaseModel, Extra

__all__ = (
    "MaybeInt",
    "MaybeFloat",
    "MaybeStr",
    "MaybePath",
    "MaybeTimedelta",
    "Model",
)

type MaybeInt = int | None
type MaybeFloat = float | None
type MaybeStr = str | None
type MaybePath = PurePath | None
type MaybeTimedelta = timedelta | None


class Model(BaseModel):
    class Config:
        extra = Extra.forbid


# noinspection PyUnresolvedReferences,PyProtectedMember
class _ModelPrinter(PrettyPrinter):
    def _pprint_model(self, obj, stream, indent, *args):
        stream.write(f"{type(obj).__name__}(")
        indent += len(type(obj).__name__) + 1
        non_nulls = {k: v for k, v in obj.__dict__.items() if v is not None}
        items = iter(non_nulls.items())
        k, v = next(items)
        stream.write(f"{k}=")
        self._format(v, stream, indent + len(k) + 1, *args)
        for k, v in items:
            stream.write(f"\n{' ' * indent}{k}=")
            self._format(v, stream, indent + len(k) + 1, *args)
        stream.write(f")")

    PrettyPrinter._dispatch[Model.__repr__] = _pprint_model
