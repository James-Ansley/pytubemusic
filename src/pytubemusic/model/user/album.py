from pydantic import Field

from .base import Model
from .cover import MaybeCover
from .tags import AlbumTags
from .track import TrackType

__all__ = ("Album",)


class Album(Model):
    """A combination of individual tracks"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: AlbumTags
    cover: MaybeCover = None
    tracks: tuple[TrackType, ...] = Field(min_length=1)
