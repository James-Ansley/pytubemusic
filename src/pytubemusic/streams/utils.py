from contextlib import AbstractContextManager, contextmanager
from typing import IO


@contextmanager
def stream[T: IO](buff: T) -> AbstractContextManager[T]:
    try:
        yield buff
    finally:
        buff.seek(0)
