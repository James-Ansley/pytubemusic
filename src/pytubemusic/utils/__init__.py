from .logs import *
from .panics import *
