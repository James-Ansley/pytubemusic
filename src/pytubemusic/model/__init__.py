import pydantic as _pydantic

from .base import *
from .album import *
from .cover import *
from .tags import *
from .track import *
from .track_data import *

type RootType = "Album | Track"
RootModel = _pydantic.RootModel[RootType]
