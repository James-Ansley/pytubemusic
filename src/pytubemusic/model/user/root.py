from pydantic import RootModel

from .album import Album
from .track import TrackType

__all__ = ("MediaType", "Media")

type MediaType = TrackType | Album


class Media:
    def __new__(cls, **kwargs) -> MediaType:
        # noinspection PyTypeChecker
        return RootModel[MediaType](**kwargs).root
