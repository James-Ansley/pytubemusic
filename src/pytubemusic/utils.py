import contextlib
import urllib.request
from collections.abc import Iterable, Mapping
from datetime import datetime, timedelta
from itertools import chain, pairwise
from pathlib import Path, PurePath
from tempfile import NamedTemporaryFile
from typing import Any

from pytubemusic.logutils import PanicOn

__all__ = [
    "to_delta", "to_timestamp", "to_microseconds", "pathify",
    "set_ends", "merge_metadata", "get_cover", "pad", "File", "TrackData",
    "StrMap", "open_or_panic",
]

StrMap = Mapping[str, Any]
TrackData = Iterable[StrMap]

File = NamedTemporaryFile


def to_delta(timestamp: str | timedelta | None) -> timedelta | None:
    """Converts a H?:M:S.f? timestamp string to a timedelta"""
    if timestamp is None:
        return None
    elif isinstance(timestamp, timedelta):
        return timestamp

    if "." in timestamp:
        timestamp, microseconds = timestamp.split(".")
    else:
        microseconds = "00"
    template = "00:00:00"
    timestamp = template[:-len(timestamp)] + timestamp + "." + microseconds
    return (
            datetime.strptime(timestamp, "%H:%M:%S.%f")
            - datetime.strptime(template, "%H:%M:%S")
    )


def to_timestamp(delta: timedelta) -> str:
    """Converts a timedelta to a H:M:S.f timestamp string"""
    return (
        f"{int(delta.total_seconds() // 3600):02.0f}"
        f":{int(delta.total_seconds() // 60) % 60:02.0f}"
        f":{int(delta.total_seconds() % 60):02.0f}"
        f".{(delta.total_seconds() % 1) * 100:02.0f}"
    )


def to_microseconds(delta: timedelta) -> int:
    """Coverts a time delta to total_microseconds"""
    return int(delta.total_seconds() * 1000)


def pathify(root: PurePath, title: str, ext=".mp3") -> PurePath:
    """
    Replaces invalid characters in filenames and joins the params to a path
    with the given extension
    """
    return root / (title.replace("/", "\u2044") + ext)


def set_ends(track_data: TrackData) -> TrackData:
    """
    Fills in missing "end" timestamps setting them to the start of the next
    track. Returns a new mapping of track data.
    """
    for t1, t2 in pairwise(chain(track_data, [{"start": None}])):
        yield t1 | {
            "start": t1.get("start"),
            "end": t1.get("end", t2["start"]),
        }


def merge_metadata(track_data: TrackData, metadata: StrMap) -> TrackData:
    """
    Merges any album metadata with track-specific metadata.
    Track metadata overrides album metadata.
    """
    for data in track_data:
        yield {**data, "metadata": {**metadata, **data["metadata"]}}


def get_cover(cover, relative_to: Path) -> "None | NamedTemporaryFile":
    """Downloads an image into a named temporary file to be used as cover art"""
    match cover:
        case {"url": url}:
            f = NamedTemporaryFile(suffix='.jpg')
            img = urllib.request.urlopen(url).read()
            f.write(img)
            return f
        case {"file": path}:
            tmp = NamedTemporaryFile(suffix='.jpg')
            msg = f"Cannot find cover file"
            with open_or_panic(relative_to.parent / path, "rb", msg) as f:
                tmp.write(f.read())
            return tmp
        case None:
            return None
        case _:
            raise ValueError(f"Unrecognised Cover format: `{cover}`")


def pad(tracks, factory, length):
    tracks = list(tracks)
    tracks += [factory() for _ in range(length - len(tracks))]
    return tracks


@contextlib.contextmanager
def open_or_panic(file_name, mode, msg):
    with (PanicOn(OSError, msg), open(file_name, mode) as f):
        yield f
