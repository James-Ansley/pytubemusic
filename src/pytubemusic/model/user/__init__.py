"""
Data types for user-entered data
"""
from .album import *
from .cover import *
from .root import *
from .tags import *
from .track import *
