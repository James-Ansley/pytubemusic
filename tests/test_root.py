from pprint import pformat

from approvaltests import verify

from pytubemusic.model.user import Merge, Playlist, Single, Split
from pytubemusic.model.user.album import Album
from pytubemusic.model.user.root import Media
from tests import test
from tests.utils import load_toml


@test(
    depends_on=(
          "test_track.py::singles_can_be_created_with_full_data",
    ),
    scope="session"
)
def root_user_model_can_parse_a_single():
    data = load_toml("single_full.toml")
    media = Media(**data)
    assert isinstance(media, Single)
    verify(pformat(media))


@test(
    depends_on=(
          "test_track.py::split_tracks_can_be_created_with_full_data",
    ),
    scope="session"
)
def root_user_model_can_parse_a_split_track():
    data = load_toml("split_track_full.toml")
    media = Media(**data)
    assert isinstance(media, Split)
    verify(pformat(media))


@test(
    depends_on=(
          "test_track.py::playlists_can_be_created_with_full_data",
    ),
    scope="session"
)
def root_user_model_can_parse_a_playlist():
    data = load_toml("playlist_full.toml")
    media = Media(**data)
    assert isinstance(media, Playlist)
    verify(pformat(media))


@test(
    depends_on=(
          "test_track.py::merge_tracks_can_be_created_with_full_data",
    ),
    scope="session"
)
def root_user_model_can_parse_a_merge_track():
    data = load_toml("merge_full.toml")
    media = Media(**data)
    assert isinstance(media, Merge)
    verify(pformat(media))


@test(
    depends_on=(
          "test_track.py::singles_can_be_created_with_full_data",
          "test_track.py::split_tracks_can_be_created_with_full_data",
          "test_track.py::playlists_can_be_created_with_full_data",
          "test_track.py::merge_tracks_can_be_created_with_full_data",
          "test_tags.py::album_tags_can_be_created_with_minimal_data",
    ),
    scope="session"
)
def root_user_model_can_parse_album():
    data = load_toml("album_full.toml")
    media = Media(**data)
    assert isinstance(media, Album)
    verify(pformat(media))
