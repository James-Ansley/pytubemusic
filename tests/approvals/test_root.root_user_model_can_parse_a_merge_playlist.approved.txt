MergePlaylist(url='www.example.com/playlist?list='
              metadata=TrackTags(title='My Track Title')
              cover=Url(href='www.example.com/pic1.jpeg')
              tracks=(TimeStub(start=datetime.timedelta(seconds=1, microseconds=500000)
                               end=datetime.timedelta(seconds=2601)),
                      Drop(drop=True),
                      TimeStub(end=datetime.timedelta(seconds=2601))))
